### DISCUSSION TITLE
picolibc using rounding mode "Round half to even" instead of "Round half up"

### DISCUSSION THREAD

**Discussion by A-Shehab** [picolibc/picolibc#742] [Q&A] [answered]
Hello everyone,

I would like to start a discussion about using "Round half to even" instead of "Round half up" in dtoa_ryu.c

```
if (vrIsTrailingZeros && lastRemovedDigit == 5 && vr % 2 == 0) {
		// Round even if the exact number is .....50..0.
		lastRemovedDigit = 4;
	}
```

in tinystdio/dtoa_ryu.c d2d function, this condition is responsible for rounding to the nearest even when the number is exactly half way between two possible rounded values.

My question is why do we use this rounding mode in picolibc instead of the conventional rounding mode which rounds the half up?

for example, 
`printf("%7.1f", 31.25);`

using picolibc this prints **31.2** however, i thought i should get **31.3** instead and that's what the online compilers i tried printed.

so what is the reason or the usage of this rounding mode in picolibc instead of the most common one?

Thanks in advance,
Best Regards,
Ahmed Shehab

**Comment by keith-packard** [comment, marked as answer]
This is what roundTiesToEven rounding attribute does in IEEE 754, when the value is exactly between two representable values, this directed rounding attribute says to round to the even value. The result is that, assuming an equal distribution of values, rounding errors tend to even out rather than having a bias towards higher values. Section 5.12 of the 2019 version of that standard says:

> Implementations shall provide conversions between each supported binary format and external decimal character sequences such that, under roundTiesToEven, conversion from the supported format to external decimal character sequence and back recovers the original floating-point representation, except that a signaling NaN might be converted to a quiet NaN. See 5.12.1 and 5.12.2 for details.

The alternative mode, roundTiesToAway, does what you suggest, but it's not the default and is not even required to be supported for a binary format implementation (like all of the floating point formats supported by picolibc).

**Comment by keith-packard** [comment]
(fwiw, this is the way that `FE_TONEAREST` rounding mode works on most (all?) current hardware for all floating point arithmetic)
